test/test.c

```diff
@@ -7170,6 +7189,67 @@ test_AuthToken(void)
     _stopServer(serverPid);
 }
 
+static const char*
+_tokenHandler(void* closure)
+{
+    return (char*) closure;
+}
+
+static void
+test_AuthTokenHandler(void)
+{
+    natsStatus          s;
+    natsConnection      *nc       = NULL;
+    natsPid             serverPid = NATS_INVALID_PID;
+    natsOptions         *opts     = NULL;
+
+    serverPid = _startServer("nats://127.0.0.1:8232", "-auth testSecret -p 8232", false);
+    CHECK_SERVER_STARTED(serverPid);
+
+    nats_Sleep(1000);
+
+    test("Connect using SetTokenHandler: ");
+    s = natsOptions_Create(&opts);
+    if (s == NATS_OK)
+        s = natsOptions_SetURL(opts, "nats://127.0.0.1:8232");
+    if (s == NATS_OK)
+        s = natsOptions_SetTokenHandler(opts, _tokenHandler, (char*) "testSecret");
+    if (s == NATS_OK)
+        s = natsConnection_Connect(&nc, opts);
+    testCond(s == NATS_OK);
+    natsConnection_Destroy(nc);
+    nc = NULL;
+
+    test("cannot set a tokenHandler when token set: ");
+    s = natsOptions_SetTokenHandler(opts, NULL, NULL);
+    if (s == NATS_OK)
+        s = natsOptions_SetToken(opts, "token");
+    if (s == NATS_OK)
+        s = natsOptions_SetTokenHandler(opts, _tokenHandler, (char*) "testSecret");
+    testCond(s == NATS_TOKEN_ALREADY_SET);
```

Comment:
Would replace those with NATS_ILLEGAL_STATE. You could (but don't have to), check that the last error we get contains the string that you expect. Again, we don't have to go there.